### PR TITLE
Update Mosviz Desktop App to Directory Parser

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

Our Mosviz notebook that powers the Mosviz Desktop App has been using our VERY outdated parsing logic (including assuming the `227.fits` cutouts!), which was preventing our current example notebook datasets from loading. After discussions with the team, we decided to move towards our directory loading structure, BUT skip out on adding any additional arguments to specify which instrument for the time being.

Under specific situations, our Mosviz parser defaults to the NIRSpec parser when a folder is given, but an unrecognized instrument is given. Because we do not have a fully generic folder parser, this PR expands on this existing paradigm and does two things:

1. Further exposes the default falling back to the NIRSpec parser to include situations where no instrument is provided at all
2. Updates the Mosviz notebook to use our directory parser, with no instrument specified (which then falls back to our NIRSpec parser)
3. STRETCH GOAL: I also took the liberty of organizing the Mosviz tests a little: a few fixtures were duplicated, so I cleaned those up. I also promoted the Mos 1D, 2D, and image fixtures to the global conftest.py in hopes that we can try to generalize them and make them broadly useful
4. Add the helper.load_data level, I explicitly checked for if the directory passed was a zip file, and if so, raise a TypeError instructing the user to unpack their data, based on direction from @orifox and @pllim 

Things still to do:

- [x] Update Documentation to specify this fallback pattern
- [x] Add additional unit tests to cover this fallback track

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/mosviz/helper.py
Comment (javerbukh) [inline jdaviz/configs/mosviz/helper.py:421]:
```suggestion
                       "NIRISS folder parsing is fully supported. Falling back to NIRSpec parsing")
```

Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/helper.py:421]: Whoops, thanks!

File: jdaviz/configs/mosviz/mosviz.ipynb
Comment (pllim) [inline jdaviz/configs/mosviz/mosviz.ipynb:23]: I am going to argue that this whole block is undesirable. Let the users uncompress data themselves. This notebook should just take a path and open stuff. No fancy pre-processing.

Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/mosviz.ipynb:23]: FYI: Reached out to @orifox who will consult with the other POs on whether we should keep or abandon this feature

Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/mosviz.ipynb]: Looks like the POs agree, so I removed that block!

File: jdaviz/conftest.py
Comment (pllim) [inline jdaviz/conftest.py:165]: Same comment as "Is random flux really necessary here?" above.

File: jdaviz/conftest.py
Comment (pllim) [inline jdaviz/conftest.py:194]: Same comment as `new_hdr` above.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:33]: Since we have complete control of `mos_image` fixture, is it possible to create a test case that is warning free? If not, the warning catching should be localized to where it is really needed to avoid accidentally ignoring warnings we should not be ignoring.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:38]:
```suggestion

    label = f"{label} 0"
    assert len(mosviz_helper.app.data_collection) == 2
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:39]:
```suggestion
    assert mosviz_helper.app.data_collection[0].label == label
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:48]: Might also want to consider data shape that isn't square so you can also catch any accidental transpose.

```suggestion
    dataval = data[label]
    assert isinstance(dataval, CCDData)
    assert dataval.shape == (55, 55)
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:64]:
```suggestion
    assert isinstance(data[labels[0]], Spectrum1D)
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:68]:
```suggestion
    spectra = [spectrum1d] * 3
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:82]:
```suggestion
    assert isinstance(data[labels[0]], Spectrum1D)
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:100]:
```suggestion
    assert data[label].shape == (1024, 15)
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:85]: Same comment as "is it possible to create a test case that is warning free?" above.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:103]: Same comment as "is it possible to create a test case that is warning free?" above.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:108]:
```suggestion
    spectra1d = [spectrum1d] * 3
    spectra2d = [mos_spectrum2d] * 3
    images = [mos_image] * 3
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:122]: Same comment as "is it possible to create a test case that is warning free?" above.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:185]: Is there a smaller dataset we can use? 80 sounds excessive. You don't want this one test to eat up all the CI quota.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:191]: This can be merged with `test_nirpsec_loader()` above using `pytest.mark.parametrized`. It is basically the same test except for that one keyword, which can be parameterized as `{'instrument': 'nirspec'}` and `{}`.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:215]:
```suggestion
def test_zip_error(mosviz_helper, tmp_path):
```

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:220]:
```suggestion
    zip_path = tmp_path / "jdaviz_test_zip.zip"
```

File: jdaviz/conftest.py
Comment (pllim) [inline jdaviz/conftest.py:210]:
```suggestion
    return CCDData(data, wcs=wcs, unit='Jy')
```

File: jdaviz/conftest.py
Comment (pllim) [inline jdaviz/conftest.py:168]:
```suggestion
    return Spectrum1D(data, wcs=wcs)
```

**Review by pllim — COMMENTED** [review]
I like the clean up. Good work!

Most comments are for the tests... 😬

File: docs/mosviz/import_data.rst
Comment (duytnguyendtn) [reply docs/mosviz/import_data.rst:32]: I'm seeing this as a double colon. Is that intended?

Comment (pllim) [reply docs/mosviz/import_data.rst:32]: Yes, that renders the code into codeblock. Currently, it is just indented. See the RTD PR build.

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:33]: I agree; I thought it was strange when I saw this warning. Unfortunately I didn't write these fixtures, and I didn't want this coding refactoring to escalate to larger than a fox hole (😉), so I didn't touch actually trying to reform these fixtures, though I agree we should spend some time refining them

Comment (pllim) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:33]: In that case, at least make the ignore more specific?

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:185]: This was one of the existing tests I migrated here. I agree, 80 is excessive, but as with my previous comment, trying to keep this to no larger than a fox hole. Definitely a candidate for future testing work slated for this quarter

Comment (pllim) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:185]: 🦊 🕳️

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py:119]: This is also one of those tests that existed prior, and that I migrated over. Same response: trying to limit to a fox hole here, though I agree with you in principle!

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py:120]: 🤷(see above comment that this was a migrated test that existed prior). I agree it's cryptic. We should fix this when we have a dedicated effort to refactor testing this quarter

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py:152]: Yep, agreed! This is definitely messy, though refactoring these tests is probably out of scope here?

File: docs/mosviz/import_data.rst
Comment (duytnguyendtn) [reply docs/mosviz/import_data.rst:32]: Ahh, whoops, missed that, thanks!

File: jdaviz/configs/mosviz/helper.py
Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/helper.py:423]: The if section above:

https://github.com/spacetelescope/jdaviz/blob/eced5e9c889f24198eff8857a7101a37312193f2/jdaviz/configs/mosviz/helper.py#L420-L424

should catch all instances where the user supplies something other than NIRISS or NIRSpec as an instrument keyword, hence an unneeded else, right? Though I suppose it could be rewritten AS an else statement if it doesn't match either case, THEN fallback? Though this way it's nice because it overrides the instrument keyword, and an unrecognized instrument follows the EXACT codepath as the NIRSpec case

Comment (pllim) [reply jdaviz/configs/mosviz/helper.py:423]: Ah, okay. I missed that. Ignore my comment on `else` then.

**Review by pllim — APPROVED** [review]
The actual implementation seems fine. If tests are 🦊 🕳️ , it is okay to defer, so I am approving now.

Maybe PO wants to approve too, especially @camipacifici who wanted this feature?

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:186]: I was super excited when you pointed that out (I love opportunities to reduce redundancy 😁), but I think the different image labels would prevent this from being parameterized, right? I would love to redesign our testing datasets so that we can parameterize the tests

**Comment by duytnguyendtn** [comment]
Whoops, one of the tests spontaneously combusted. Will get to that in a bit

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:186]: You can parametrize more than one args at a time. See https://docs.pytest.org/en/stable/how-to/parametrize.html .

But I didn't realize most of these test codes are pre-existing, so I'll leave it up to you whether you want to address this or not.

**Comment by camipacifici** [comment]
I loaded the nirspec level3 dataset with
`jdaviz --layout=mosviz mosviz_nirspec_data_0.3/level3` and it loads, but with a few yellow snackbar messages.
At loading, the first 2D spectrum does not show automatically, but it appears if I click on the second object and on the first object again. Any idea why?
Screen Shot 2022-04-08 at 10 24 05 PM
Screen Shot 2022-04-08 at 10 24 19 PM

Loading the NIRISS dataset is out of scope here? If so, a new ticket is needed.

**Comment by duytnguyendtn** [comment]
> the first 2D spectrum does not show automatically

@camipacifici If you load the same dataset via notebook and manually specify `instrument='nirspec'` into `load_data`, do you get the same behavior?

**Comment by duytnguyendtn** [comment]
> Loading the NIRISS dataset is out of scope here?

Yep, that's correct. I was instructed explicitly to NOT add a new argument to the command line to allow the user to specify which instrument at this time, so it defaults back to nirspec, which is our global loader. This is actually the reason for 

> a few yellow snackbar messages

because it's effectively falling back, since we can't actually specify which instrument via the command line right now

**Comment by camipacifici** [comment]
> If you load the same dataset via notebook and manually specify `instrument='nirspec'` into `load_data`, do you get the same behavior?

In notebook, if I load without specifying `instrument='nirspec'` the first 2D spectrum is there.

**Comment by camipacifici** [comment]
> I was instructed explicitly at this time to NOT add a new argument to the command line

Thank you. Totally fine, but please make a new ticket to support NIRISS/NIRCam data as well in the future.

**Comment by duytnguyendtn** [comment]
> In notebook, if I load without specifying `instrument='nirspec'` the first 2D spectrum is there.

Very strange! I can reproduce this, but I have NO idea why that this would happen, must investigate; thanks for catching this!

**Comment by duytnguyendtn** [comment]
Investigated the issue above reported by @camipacifici. Mix of red herrings and confusing threads, but here it is:

1. I was able to reproduce the error in the notebook interface as well, so this issue isn't just present in desktop app. Thus, the desktop app doesn't seem to be the issue. It's MUCH easier to reproduce in desktop app mode, but it's possible to hit it in both contexts/modes.
2. The data is actually there. I spun my wheels a bit investigating why the data wasn't being added properly, but it's actually there. You don't even need to deselect or reselect the data option to make it show up. If I interact with the pan tool, the data pops up. 

I'll bring it up with the SMEs tomorrow. Maybe @mariobuikhuizen has some ideas about how to even begin debugging this one?

**Comment by kecnry** [comment]
> the first 2D spectrum does not show automatically

I've noticed this in the past when working on mosviz and it seems to happen somewhat randomly.  My guess is that signals are not being processed in a consistent order, so sometimes the data is not shown correctly.  Its definitely something we should look into in more detail and try to fix once and for all though (but out of scope here)!

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (kecnry) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:21]: this probably shouldn't be needed (may be related to @camipacifici's comment - so having in the test might just help to ensure everything is set correct).  There is also a higher API for this now (`table.select_row(0)`), but that is essentially just a wrapper around this.

(I also realize these were here before and you just moved them - so making changes in likely out of scope - I just wanted to leave a mention somewhere in case it helps us when debugging the spectrum showing on load issue).

**Review by kecnry — APPROVED** [review]
Looks good to me!  I agree that a lot of the testing comments should be addressed eventually but can be deferred here since the tests were just moved and aren't introduced by this PR.

**Comment by duytnguyendtn** [comment]
Thanks for the context @kecnry! I'll file a ticket for your bug Cami refound before I merge. Does this sound OK @camipacifici?

**Comment by camipacifici** [comment]
> Thanks for the context @kecnry! I'll file a ticket for your bug Cami refound before I merge. Does this sound OK @camipacifici?

yes

**Comment by duytnguyendtn** [comment]
Thanks everyone for the reviews!

**Pull request merged**
